### ISSUE TITLE
Enum withPredicate

### ISSUE THREAD

**Issue by Confidenceman02** [davidmdm/myzod#57]
First of all, what an excellent project and thank you very much for the hard work.

I am building an ORM for redis using myzod types as a utility for redis schemas and have run in to an interesting case caused by both the weirdness of enums and the way redis stores values in hash objects.

Given the enum and enum schema:

```javascript
enum Scope {
  Hello,
  World
}

mz.enum(Scope)
```
The value `Scope.Hello` will be stored in redis as the string "0" instead of the number 0. On retrival and parsing this will throw as expected:

```javascript
await redis.storeTheScope(Scope.Hello) // 0
const scopeValue = await redis.getTheScope() // "0"
mz.enum(Scope).parse(scopeValue) // throws
```
Whilst myzod is absolutely doing the right thing here, in order to not allow numerical enums in my library I was surprised to see `EnumType` does not contain a `withPredicate` method like `StringType`.

I am happy to do the work to make this happen but wanted to ask if this was something you would consider in the library?

Regards

**Comment by davidmdm** [comment]
Hi @Confidenceman02 !

The reason that enumtype does not have a withPredicate function is because there's usually no good reason to limit the value of an enum at runtime.

Let's stay that for a string, you are writing an API that needs to validate a field as a string, but that string needs to be a semver version, you can can reduce the type string to any string that matches a semver regexp.

```ts
const semverSchema = z.string().withPredicate(
   value => /^v\d+.\d+\d+$/.match(value), 
   'value must be a semver string'
)
``` 

However for an enum, if you want to validate against a subset of the enum, it is more useful to declare the subset enum at compile time.

Regardless, I believe what you actually are searching for is the literals schema type.

```ts
const schema = z.literals('hello', 'world')

type MyType = z.Infer<typeof schema>; // 'hello' | 'world'
```

This will make sure that your values must match the literal values you provided to the schema exactly.


Is this satisfactory or is withPredicate on schema still desirable for you?

**Comment by Confidenceman02** [comment]
Many thanks for the response!

I do see your point and agree with you on why `withPredicate` is missing on enums. Limiting the enum via `withPredicate` was not the right fit with what I actually need.

In a nutshell what I need is access to the underlying enum passed to the schema which `mz.enum(Scope)` makes opaque.
The reason I need this is for mapping values back in to their enum representation.

Consider the following enum:
```javascript
enum Scope {
  Hello,
  World
}
```
At runtime this ends up looking like:
```javascript
{
  "0": "Hello",
  "1": "World",
  "Hello": 0,
  "World": 1
}
```

As mentioned previously a value such as `Scope.Hello` which is actually a number value of `0` in this scenario, ends up being a string in redis `"0"`. When retrieving this value, which I know belongs to the enum Scope, and passing it to the enum schema it will throw as intended, because `"0"` does not exist in the enum values.

What I need is a way to get the value `"0"` and convert it to its enum representation `0`. As you can see above, the runtime enum gives me the ability to do just that as it is a reverse map.

From a psuedo code perspective, here is the extent of what I would need.

```javascript
// Let's say this schema was generated by a user >>>>
enum Scope {
  Hello,
  World
}
const enumSchema = mz.enum(Scope) 

// Lets say at some point a user persist a scope value
await redis.storeTheScope(Scope.Hello) // 0 gets turns in to "0" by redis

// Now the user wants to retrieve the scope value
const scopeValue = await redis.getTheScope() // "0"

// I know this value belongs to a Scope enum, but I don't know it its meant to be a string or a number.
// so I  map the value to its enum representation.
const enumMap = enumSchema.enumShape //Myzod gives me the runtime enum object
const mappedValue = enumMap[enumMap[scopeValue]] // 0

// Validate
enumSchema.parse(mappedValue) // validates!
```
From a myzod perspective all it needs to do is expose the underlying enum that is passed to it. With that I can do what I need to do. Because enums in javascript are actually data, the schema exposing it is actually very helpful.

Another possible solution is for myzod to do this mapping for you which might avoid a current issue with the `EnumType` `parse` method.

Currently the `EnumType` class strips the values from the runtime enum, so for the above example the values would be

```javascript
["Hello", "World", 0, 1]
```
The `parse` method checks a given value exists in the enum values.

```javascript
// myzod EnumType code
if (!this.values.includes(coercedValue)) {
      throw this.typeError(`error ${JSON.stringify(value)} not part of enum values`);
    }
```
This means  that myzod will validate `Scope` with the following code.

```javascript
const with0 = mz.enum(Scope).parse(0) // validates
const with1 = mz.enum(Scope).parse(1) // validates
const withHello = mz.enum(Scope).parse("Hello") // validates
const withWorld = mz.enum(Scope).parse("World") // validates


Scope.Hello === with0 // true
Scope.World === with1 // true
Scope.Hello === withHello // false
Scope.World === withWorld // false 
```
It is entirely possible people will do this enum check with a validated myzod value. In my mind, if myzod validates a value then the following should be true which currently is not.

```javascript
const validatedvalue = mz.enum(Scope).parse("Hello") // validates

validatedValue === Scope.Hello || validatedValue === Scope.World // should be true but is not
``` 

Sorry for the long winded journey, keen to hear what you think. All things being said if the enum value was exposed I would be a happy camper. :)

Regards

**Comment by davidmdm** [comment]
Bewarned, I am going to push back a little, but know that at the end of the day, I am not opposed to adding the enumShape to EnumType. I just want to see if we can solve your problem without changing anything to myzod.

First let's talk about enums. _**enums in TS are weird**_.

In fact they are especially weird if you write them like in your example:
```ts
enum Scope {
  Hello,
  World
}
console.log(JSON.stringify(Scope)) // {"0":"Hello","1":"World","Hello":0,"World":1}
```

However if you write it like so, most of your problems will go away, at the cost of a couple more keystrokes:
```ts
enum Scope {
  Hello = "hello",
  World = "world
}

console.log(JSON.stringify(Scope)) // {"Hello":"hello","World":"world"}
```

In which case:

```ts

const schema = z.enum(Scope)

const hello = schema.parse('hello') // "hello"

console.log(hello === Scope.Hello) // true
```

I believe this would be easier than trying to map back and forth using the first type of enum we saw.
It would also solve the problem of writing the string "0" to redis which is not what you want per se.


As for exposing the enum, I have done it for other types, and I am open to doing it again for enums, however you technically should already have the type. For example:

`scopes.ts`
```ts
export enum Scope {
  Hello = "hello",
  World = "world"
}

export const scopeSchema = z.enum(Scope)
```

`example.ts`
```ts
import {Scope, scopeSchema} from '~/scopes`
```

Now this being said, I aim to please. So if this hasn't convinced you or given you an idea for a better way, and you desire the enum to be exposed from the enumType, give me the word and I will add it!

Hope this helps :)

**Comment by Confidenceman02** [comment]
Always appreciate some push back! :)

Indeed they are weird! Especially with enums such as:
```javascript
enum Scope {
  Hello = 4554
  World = "djsdjshdjshd"
}
```
In regards to:
> However if you write it like so, most of your problems will go away, at the cost of a couple more keystrokes:

You are completely right, however as I mentioned in the comment in one of the examples, the schemas are user defined, so although I used the example as if I was in command of the EnumType and the enum itself, I actually am not. This means any EnumType schema that a user creates, in all the lovely weird enum variations could show up in my program.

This is why my first intuition was to be able set the `withPredicate` on the passed in enum schema to protect against numerical enums. I am confident this is the wrong method though. Having access to the underlying runtime enum is actually a much better solution because it allows me to not limit the enum values and instead map any redis string transformations to something consistent that will validate.

>Now this being said, I aim to please. So if this hasn't convinced you or given you an idea for a better way, and you desire the enum to be exposed from the enumType, give me the word and I will add it!

I massively appreciate your flexibility with this, and your efforts in helping find a solution that protects myzod against changes. It's how I would do things also!

That being said, if you feel it is a worthy request and are kind enough to add it, I would be delighted!

Many thanks

**Comment by davidmdm** [comment]
@Confidenceman02 

I released a new version `1.10.0` which adds support for `enum()` and `withPredicate` to EnumType:

```ts
const schema = z.enum(Colors)
schema.enum() === Colors // true

// ---

const schema = z.enum(Colors)
   .withPredicate(value => typeof value === 'string', 'color cannot be numeric');
```

It is not in the readme yet but I will add it there soon. Let me know if this issue is resolved!

**Comment by Confidenceman02** [comment]
Wonderful, and yes! Solves my problem nicely. 

Love the `withPredicate` example also 👍
